src/System/Console/GetOpt/Generics.hs

```diff
@@ -118,17 +135,22 @@ mkEmptyArguments fields = case (sing :: Sing xs, fields) of
 -- * showing help?
 
 helpWrapper :: (All Option xs) =>
-  [String] -> NP FieldInfo xs -> a -> Either (IO ()) a
-helpWrapper args fields a =
+  String -> [Hint] -> [String] -> NP FieldInfo xs -> Either [String] a -> Result a
+helpWrapper header hints args fields result =
     case getOpt Permute [helpOption] args of
-      ([], _, _) -> Right a
-      (() : _, _, _) -> Left $ do
-        progName <- getProgName
-        let header = progName
-        putStrLn (usageInfo header (mkOptDescrs fields))
+      ([], _, _) -> case result of
```

Comment:
@jkarni: `helpWrapper` finds out, whether `--help` was given or not. If not it converts `result` to type `result`. If yes it returns `OutputAndExit` with the help message.

To be honest I'm not sure if you were asking about `helpWrapper`. Were you? How can we make the code clearer?